### PR TITLE
Validation specs

### PR DESCRIPTION
**Pull request by MarcoDDM**
**Description**

This PR focuses on improving model validation and adding RSpec unit tests for the `Comment`, `Like`, `Post`, and `User `models.

**Changes**
- Strengthened model validation logic.
- Added comprehensive RSpec unit tests.
- Testing
- All changes are rigorously tested using RSpec to ensure accurate model behavior.
- Added screenshot for code reviewer's convenience on the testing aspect.
rspec_blog-app

Thanks for your time and review. Don't hesitate to contact me if there are any changes needed.

### CODE REVIEW

File: spec/models/user_spec.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+RSpec.describe User, type: :model do
+  it 'is invalid without a name' do
+    user = User.new(posts_counter: 0)
+    expect(user).to_not be_valid
+  end
+
+  it 'is invalid with a negative posts_counter' do
+    user = User.new(name: 'Tom', posts_counter: -1)
+    expect(user).to_not be_valid
+  end
```
Comment (topeogunleye) [inline spec/models/user_spec.rb:10]: ---

### Positive Feedback:
It's fantastic to see that you're actively validating critical attributes within the `User` model, such as ensuring that the `posts_counter` cannot be negative. Well done on implementing this test—it's a key part of maintaining data integrity in the application.

### Room for Improvement:
While you've effectively tested for the absence of a name and negative values of `posts_counter`, I noticed that there are no tests checking `posts_counter` when it has a value of zero or is a positive number. These test cases are important to confirm that the system treats valid input as expected.

```.txt
PostsCounter must be an integer greater than or equal to zero.
```

### Suggestion for Improvement:
Let's round out your test coverage by including scenarios where `posts_counter` is set to zero, which should be valid, as well as when it's positive. Here’s an example of how these test cases might look:

```ruby
it 'is valid with a zero posts_counter' do
  user = User.new(name: 'Tom', posts_counter: 0)
  expect(user).to be_valid
end

it 'is valid with a positive posts_counter' do
  user = User.new(name: 'Tom', posts_counter: 1)
  expect(user).to be_valid
end
```

Adding these tests will help ensure that your validations are not too restrictive and are allowing legitimate cases.

### Final Note:
Great work on the testing you've done so far. By expanding your tests to cover these additional cases, you'll be ensuring that your model validations are comprehensive and robust. Thorough testing is a hallmark of reliable and maintainable software.

---

File: spec/models/post_spec.rb [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    post = Post.new(
+      title: 'Title',
+      comments_counter: -1,
+      likes_counter: 0,
+      author: tom
+    )
+    expect(post).to_not be_valid
+  end
+
+  it 'is invalid with a negative likes_counter' do
```
Comment (topeogunleye) [inline spec/models/post_spec.rb:46]: Good job on your test coverage! Your attention to the required validations in the Post model is commendable. You've successfully written tests to check for the presence of a title, character limits, and author association, which are crucial for the integrity of the Post model.

### Room for Improvement:
However, I noticed that while we have tests for negative values for `comments_counter` and `likes_counter`, we don't have explicit tests for zero and positive values. It's essential to test the boundary conditions of our validations to make sure they behave correctly across the expected range of inputs.
```.txt
LikesCounter must be an integer greater than or equal to zero.
```
### Suggestion for Improvement:
Consider adding test cases for zero and positive values to ensure comprehensive coverage. For instance, a test for `likes_counter` could be as follows:

```ruby
it 'is valid with a zero likes_counter value' do
  post = Post.new(
    title: 'Valid Title',
    comments_counter: 0,
    likes_counter: 0,
    author: tom
  )
  expect(post).to be_valid
end

it 'is valid with a positive likes_counter value' do
  post = Post.new(
    title: 'Valid Title',
    comments_counter: 0,
    likes_counter: 1,
    author: tom
  )
  expect(post).to be_valid
end
```

Adding such tests will help in confirming that our model not only rejects invalid data but also accepts valid data as it should. This is just as important in TDD (Test-Driven Development) to prevent any false positives and to ensure the robustness of your application.

### Final Note:
Again, great work on your tests so far. By expanding the test cases to cover the full range of valid and invalid inputs, you will further solidify your application's reliability and maintainability.

Keep it up!

---

File: spec/models/post_spec.rb [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    post = Post.new(
+      title:,
+      comments_counter: 0,
+      likes_counter: 0,
+      author: tom
+    )
+    expect(post).to_not be_valid
+  end
+
+  it 'is invalid with a negative comments_counter' do
```
Comment (topeogunleye) [inline spec/models/post_spec.rb:36]: You're making great progress with your tests, and I can see that you've put effort into verifying the `Post` model behaves correctly under various conditions. Testing for a negative `comments_counter` is crucial and you've implemented this test case well.

### Room for Improvement:
However, we should also verify that the `comments_counter` behaves correctly when it has a zero or a positive value. Currently, there's no explicit test that confirms the model's validity when `comments_counter` is within this acceptable range.

### Suggestion for Improvement:
To ensure our model validation is working as expected across all scenarios, we should include tests for zero and positive values for the `comments_counter`. Here is how you could write these test cases:

```ruby
it 'is valid with a zero comments_counter value' do
  post = Post.new(
    title: 'Title',
    comments_counter: 0,
    likes_counter: 0,
    author: tom
  )
  expect(post).to be_valid
end

it 'is valid with a positive comments_counter value' do
  post = Post.new(
    title: 'Title',
    comments_counter: 1,
    likes_counter: 0,
    author: tom
  )
  expect(post).to be_valid
end
```

These test cases will help ensure that the `Post` model validations are not only rejecting negative values but also properly accepting values that should be valid.

### Final Note:
Keep up the excellent work; by adding these additional test cases, you will have a more robust and reliable suite of tests. Ensuring comprehensive test coverage is an excellent practice for maintaining high code quality.

---

**Review by topeogunleye — CHANGES_REQUESTED** [review]
Hi @MarcoDDM, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

Try-Again

> ## (Highlights) Good Points: 💪 

---

### Well-Implemented Aspects:

1. **Negative Value Handling**: You've done an excellent job ensuring that your model validations handle negative numbers. The test you've written to check that a negative `posts_counter` is invalid is spot on, and this is critical for preventing invalid data from entering your system.

2. **Presence Validation for Name**: You've correctly identified and tested that a user should not be valid without a name. This is a fundamental validation that helps maintain data quality in the application.

3. **Test Clarity**: Your tests are clear and concise, which makes them easy to understand and maintain. This clarity in testing is essential for you and anyone else who might work on this codebase in the future.

4. **Focused Test Cases**: Each of your tests focuses on a single piece of functionality. This focus is a good testing practice because it means that when a test fails, it's easy to pinpoint exactly what went wrong.

5. **Adherence to Requirements**: From the tests provided, it appears you have a good understanding of the requirements and have translated them into specific test cases. This is a great approach to ensure that all the specified requirements are met.

### Encouragements:

Keep up the focused approach to testing individual requirements. As you add more tests, you'll build up a robust suite that will guard against regressions and ensure that your application behaves as expected now and as it continues to evolve.

--- 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- x
- y
- z

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag @topeogunleye in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/models/post_spec.rb (hunk 3 above)
Comment (MarcoDDM) [reply spec/models/post_spec.rb:36]: Thanks for your hints. I will apply all you said. This was really helpful for me.

**Review by anagudelogu — APPROVED** [review]
# STATUS: APPROVED! 🍾 🎆 🥇 

Hi @MarcoDDM, 
I'm Andres.

You did a great job working on this milestone, your project is complete! 🚀 
There is nothing else to say other than... it's time to merge it 🤖  
Congratulations! 🎉 



## Highlights
✅ Great use of Git flow. 💯 
✅ Your project is documented professionally. 📜 
✅ Descriptive PR title, summary and commit messages. 💬 

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
